File: src/main/java/fr/traqueur/energylib/hooks/EnergyItemsAdderCompatibility.java
package fr.traqueur.energylib.hooks;

import dev.lone.itemsadder.api.Events.CustomBlockBreakEvent;
import dev.lone.itemsadder.api.Events.CustomBlockPlaceEvent;
import fr.traqueur.energylib.EnergyLib;
import fr.traqueur.energylib.api.EnergyAPI;
import fr.traqueur.energylib.api.EnergyManager;
import fr.traqueur.energylib.api.exceptions.SameEnergyTypeException;
import io.th0rgal.oraxen.api.events.noteblock.OraxenNoteBlockBreakEvent;
import io.th0rgal.oraxen.api.events.noteblock.OraxenNoteBlockPlaceEvent;
import io.th0rgal.oraxen.compatibilities.CompatibilityProvider;
import org.bukkit.Location;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.inventory.ItemStack;

/**
 * This class is a compatibility for ItemsAdder plugin.
 * It allows to place and break energy components from ItemsAdder.
 */
public class EnergyItemsAdderCompatibility implements Listener {

    /**
     * The EnergyAPI instance.
     */
    private final EnergyAPI api;

    /**
     * The EnergyManager instance.
     */
    private final EnergyManager energyManager;

    /**
     * Constructor.
     * @param api The EnergyAPI instance.
     */
    public EnergyItemsAdderCompatibility(EnergyAPI api) {
        this.api = api;
        this.energyManager = api.getManager();
    }

    /**
     * Handle the place of a custom block.
     * @param event The event.
     */
    @EventHandler
    public void onPlace(CustomBlockPlaceEvent event) {
        ItemStack item = event.getItemInHand();
        if (!energyManager.isComponent(item)) {
            return;
        }
        Location location = event.getBlock().getLocation();
        api.getScheduler().runAsync((t) -> {
            var component = energyManager.createComponent(item);
            try {
                energyManager.placeComponent(component, location);
            } catch (SameEnergyTypeException ignored) {}
        });
    }

    /**
     * Handle the break of a custom block.
     * @param event The event.
     */
    @EventHandler
    public void onBreak(CustomBlockBreakEvent event) {
        Location location = event.getBlock().getLocation();
        if (!energyManager.isBlockComponent(location)) {
            return;
        }
        api.getScheduler().runAsync((t) -> energyManager.breakComponent(location));
    }

}

File: src/main/java/fr/traqueur/energylib/hooks/EnergyItemsAdderCompatibility.kt
/*package fr.traqueur.energylib.hooks

import fr.traqueur.energylib.api.EnergyAPI
import fr.traqueur.energylib.api.EnergyManager
import fr.traqueur.energylib.api.exceptions.SameEnergyTypeException
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent

/**
 * This class is a compatibility for ItemsAdder plugin.
 * It allows to place and break energy components from ItemsAdder.
 */
class EnergyItemsAdderCompatibility(
    /**
     * The EnergyAPI instance.
     */
    private val api: EnergyAPI
) : Listener {

    /**
     * The EnergyManager instance.
     */
    private val energyManager: EnergyManager = api.manager!!

    /**
     * Handle the place of a custom block.
     * @param event The event.
     */
    @EventHandler
    fun onPlace(event: BlockPlaceEvent) {
        val item = event.getItemInHand().clone()
        if (!energyManager.isComponent(item)) {
            return
        }
        val location = event.getBlock().location
        api.scheduler?.runAtLocation(location, { t ->
            val component = energyManager.createComponent(item)
            try {
                energyManager.placeComponent(component, location)
            } catch (ignored: SameEnergyTypeException) {
            }
        })
    }

    /**
     * Handle the break of a custom block.
     * @param event The event.
     */
    @EventHandler
    fun onBreak(event: BlockBreakEvent) {
        val location = event.getBlock().location
        if (!energyManager.isBlockComponent(location)) {
            return
        }
        event.isCancelled = true
        api.scheduler?.runAtLocation(location, { t -> energyManager.breakComponent(event.getPlayer(), location) })
    }
}/

 */
